### PR TITLE
Add benchmarks for on_initial_header

### PR DESCRIPTION
**Pull request by hcaseyal**
Currently, the on_initial_header function is not microbenchmarked in the hpack benchmarks. This is problematic because other tools have shown on_initial_header has significant overhead on the hpack parse path and we need these numbers.

In the long run, it would be better to properly mock out the transport so we can directly call the on_initial_header function instead of doing the copy and paste thing that I did. I will take this as an action item for next quarter. For now, this change is better than not having this code path microbenchmarked at all (and note that the existing microbenchmark called OnHeaderNew was a copy paste thing, so this is certainly not worse than the existing microbenchmark).

This PR adds new hpack benchmarks to benchmark on_initial_header and renames an existing benchmark, OnHeaderNew, to a more representative name, OnHeaderTimeout. The overhead of re-creating the arena every 4000 iterations adds a constant ~200ns to each HpackParserParseHeader benchmark.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,099      Total (>)      1,961,059

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,883,606      Total (<)     10,883,611

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,059      Total (=)      1,961,059

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,883,606      Total (<)     10,883,611

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```


Missing files (indicates new benchmark): 
    bm_chttp2_hpack.BM_HpackParserParseHeader_SameDeadline_OnHeaderNew_.opt.new: 10
    bm_chttp2_hpack.BM_HpackParserParseHeader_SameDeadline_OnHeaderNew_.counters.new: 10


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```


Missing files (indicates new benchmark): 
    bm_chttp2_hpack.BM_HpackParserParseHeader_SameDeadline_OnHeaderNew_.opt.new: 10
    bm_chttp2_hpack.BM_HpackParserParseHeader_SameDeadline_OnHeaderNew_.counters.new: 10


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,059      Total (=)      1,961,059

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,883,613      Total (>)     10,883,607

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```


Missing files (indicates new benchmark): 
    bm_chttp2_hpack.BM_HpackParserParseHeader_SameDeadline_OnHeaderNew_.opt.new: 10
    bm_chttp2_hpack.BM_HpackParserParseHeader_SameDeadline_OnHeaderNew_.counters.new: 10


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,059      Total (=)      1,961,059

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,883,614      Total (>)     10,883,612

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,059      Total (=)      1,961,059

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,883,607      Total (<)     10,883,612

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```


Missing files (indicates new benchmark): 
    bm_chttp2_hpack.BM_HpackParserParseHeader_SameDeadline_OnHeaderNew_.opt.new: 10
    bm_chttp2_hpack.BM_HpackParserParseHeader_SameDeadline_OnHeaderNew_.counters.new: 10


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,099      Total (=)      1,961,099

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,883,605      Total (<)     10,883,613

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```


Missing files (indicates new benchmark): 
    bm_chttp2_hpack.BM_HpackParserParseHeader_SameDeadline_OnHeaderNew_.opt.new: 10
    bm_chttp2_hpack.BM_HpackParserParseHeader_SameDeadline_OnHeaderNew_.counters.new: 10


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,099      Total (=)      1,961,099

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,883,610      Total (<)     10,883,612

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```


Missing files (indicates new benchmark): 
    bm_chttp2_hpack.BM_HpackParserParseHeader_SameDeadline_OnHeaderNew_.opt.new: 10
    bm_chttp2_hpack.BM_HpackParserParseHeader_SameDeadline_OnHeaderNew_.counters.new: 10


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,099      Total (=)      1,961,099

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,883,602      Total (<)     10,883,618

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```


Missing files (indicates new benchmark): 
    bm_chttp2_hpack.BM_HpackParserParseHeader_SameDeadline_OnHeaderNew_.opt.new: 10
    bm_chttp2_hpack.BM_HpackParserParseHeader_SameDeadline_OnHeaderNew_.counters.new: 10


[microbenchmarks] Performance differences noted:
Benchmark                                                                        cpu_time    real_time
-------------------------------------------------------------------------------  ----------  -----------
BM_HpackParserParseHeader<AddIndexedSingleInternedElem, UnrefHeader>             +100%       +99%
BM_HpackParserParseHeader<AddIndexedSingleStaticElem, UnrefHeader>               +151%       +151%
BM_HpackParserParseHeader<EmptyBatch, UnrefHeader>                               +1806%      +1806%
BM_HpackParserParseHeader<IndexedSingleInternedElem, UnrefHeader>                +328%       +328%
BM_HpackParserParseHeader<IndexedSingleStaticElem, UnrefHeader>                  +742%       +742%
BM_HpackParserParseHeader<KeyIndexedSingleInternedElem, UnrefHeader>             +122%       +122%
BM_HpackParserParseHeader<KeyIndexedSingleStaticElem, UnrefHeader>               +89%        +89%
BM_HpackParserParseHeader<MoreRepresentativeClientInitialMetadata, UnrefHeader>  +27%        +27%
BM_HpackParserParseHeader<NonIndexedBinaryElem<1, false>, UnrefHeader>           +83%        +83%
BM_HpackParserParseHeader<NonIndexedBinaryElem<1, true>, UnrefHeader>            +86%        +86%
BM_HpackParserParseHeader<NonIndexedBinaryElem<10, false>, UnrefHeader>          +30%        +30%
BM_HpackParserParseHeader<NonIndexedBinaryElem<10, true>, UnrefHeader>           +92%        +92%
BM_HpackParserParseHeader<NonIndexedBinaryElem<100, true>, UnrefHeader>          +62%        +62%
BM_HpackParserParseHeader<NonIndexedBinaryElem<3, false>, UnrefHeader>           +51%        +51%
BM_HpackParserParseHeader<NonIndexedBinaryElem<3, true>, UnrefHeader>            +96%        +96%
BM_HpackParserParseHeader<NonIndexedBinaryElem<31, false>, UnrefHeader>          +31%        +31%
BM_HpackParserParseHeader<NonIndexedBinaryElem<31, true>, UnrefHeader>           +61%        +61%
BM_HpackParserParseHeader<NonIndexedElem, UnrefHeader>                           +113%       +113%
BM_HpackParserParseHeader<RepresentativeClientInitialMetadata, UnrefHeader>      +125%       +125%
BM_HpackParserParseHeader<RepresentativeServerInitialMetadata, UnrefHeader>      +476%       +476%
BM_HpackParserParseHeader<RepresentativeServerTrailingMetadata, UnrefHeader>     +377%       +377%
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,099      Total (=)      1,961,099

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,883,611      Total (<)     10,883,617

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,297      Total (=)      1,961,297

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,884,268      Total (>)     10,884,262

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```


Missing files (indicates new benchmark): 
    bm_chttp2_hpack.BM_HpackParserParseHeader_SameDeadline_OnHeaderNew_.opt.new: 10
    bm_chttp2_hpack.BM_HpackParserParseHeader_SameDeadline_OnHeaderNew_.counters.new: 10


[microbenchmarks] Performance differences noted:
Benchmark                                                                        cpu_time    real_time
-------------------------------------------------------------------------------  ----------  -----------
BM_HpackParserParseHeader<AddIndexedSingleInternedElem, UnrefHeader>             +70%        +70%
BM_HpackParserParseHeader<AddIndexedSingleStaticElem, UnrefHeader>               +141%       +141%
BM_HpackParserParseHeader<IndexedSingleInternedElem, UnrefHeader>                +166%       +166%
BM_HpackParserParseHeader<IndexedSingleStaticElem, UnrefHeader>                  +128%       +128%
BM_HpackParserParseHeader<KeyIndexedSingleInternedElem, UnrefHeader>             +108%       +108%
BM_HpackParserParseHeader<KeyIndexedSingleStaticElem, UnrefHeader>               +31%        +31%
BM_HpackParserParseHeader<MoreRepresentativeClientInitialMetadata, UnrefHeader>  +37%        +37%
BM_HpackParserParseHeader<NonIndexedBinaryElem<1, false>, UnrefHeader>           +88%        +88%
BM_HpackParserParseHeader<NonIndexedBinaryElem<1, true>, UnrefHeader>            +109%       +109%
BM_HpackParserParseHeader<NonIndexedBinaryElem<10, false>, UnrefHeader>          +24%        +24%
BM_HpackParserParseHeader<NonIndexedBinaryElem<10, true>, UnrefHeader>           +36%        +36%
BM_HpackParserParseHeader<NonIndexedBinaryElem<100, false>, UnrefHeader>         +8%         +8%
BM_HpackParserParseHeader<NonIndexedBinaryElem<100, true>, UnrefHeader>          +71%        +71%
BM_HpackParserParseHeader<NonIndexedBinaryElem<3, false>, UnrefHeader>           +47%        +47%
BM_HpackParserParseHeader<NonIndexedBinaryElem<3, true>, UnrefHeader>            +39%        +39%
BM_HpackParserParseHeader<NonIndexedBinaryElem<31, false>, UnrefHeader>          +45%        +45%
BM_HpackParserParseHeader<NonIndexedBinaryElem<31, true>, UnrefHeader>           +101%       +101%
BM_HpackParserParseHeader<NonIndexedElem, UnrefHeader>                           +66%        +66%
BM_HpackParserParseHeader<RepresentativeClientInitialMetadata, UnrefHeader>      +90%        +90%
BM_HpackParserParseHeader<RepresentativeServerInitialMetadata, UnrefHeader>      +259%       +259%
BM_HpackParserParseHeader<RepresentativeServerTrailingMetadata, UnrefHeader>     +91%        +91%
```

File: test/cpp/microbenchmarks/bm_chttp2_hpack.cc [hunk 1]
```diff
@@ -766,8 +775,57 @@ class RepresentativeServerTrailingMetadata {
 
 static void free_timeout(void* p) { gpr_free(p); }
 
-// New implementation.
-static void OnHeaderNew(void* user_data, grpc_mdelem md) {
+// Benchmark the current on_initial_header implementation
+static void OnInitialHeader(void* user_data, grpc_mdelem md) {
+  // Setup for benchmark. This will bloat the absolute values of this benchmark
+  grpc_chttp2_incoming_metadata_buffer buffer;
+  grpc_chttp2_incoming_metadata_buffer_init(&buffer, (gpr_arena*)user_data);
```
Comment (yang-g) [inline test/cpp/microbenchmarks/bm_chttp2_hpack.cc]: Do you need to destroy the buffer?

Comment (hcaseyal) [reply test/cpp/microbenchmarks/bm_chttp2_hpack.cc]: I probably do :)

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,241      Total (=)      1,961,241

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,884,146      Total (<)     10,884,150

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```


Missing files (indicates new benchmark): 
    bm_chttp2_hpack.BM_HpackParserParseHeader_SameDeadline_OnHeaderNew_.opt.new: 10
    bm_chttp2_hpack.BM_HpackParserParseHeader_SameDeadline_OnHeaderNew_.counters.new: 10


[microbenchmarks] Performance differences noted:
Benchmark                                                                        cpu_time    real_time
-------------------------------------------------------------------------------  ----------  -----------
BM_HpackParserParseHeader<AddIndexedSingleInternedElem, UnrefHeader>             +82%        +82%
BM_HpackParserParseHeader<AddIndexedSingleStaticElem, UnrefHeader>               +133%       +133%
BM_HpackParserParseHeader<EmptyBatch, UnrefHeader>                               +529%       +529%
BM_HpackParserParseHeader<MoreRepresentativeClientInitialMetadata, UnrefHeader>  +6%         +6%
BM_HpackParserParseHeader<NonIndexedBinaryElem<1, false>, UnrefHeader>           +72%        +72%
BM_HpackParserParseHeader<NonIndexedBinaryElem<1, true>, UnrefHeader>            +114%       +114%
BM_HpackParserParseHeader<NonIndexedBinaryElem<10, false>, UnrefHeader>          +72%        +72%
BM_HpackParserParseHeader<NonIndexedBinaryElem<10, true>, UnrefHeader>           +121%       +121%
BM_HpackParserParseHeader<NonIndexedBinaryElem<100, false>, UnrefHeader>         +6%         +6%
BM_HpackParserParseHeader<NonIndexedBinaryElem<100, true>, UnrefHeader>          +58%        +58%
BM_HpackParserParseHeader<NonIndexedBinaryElem<3, false>, UnrefHeader>           +57%        +57%
BM_HpackParserParseHeader<NonIndexedBinaryElem<3, true>, UnrefHeader>            +99%        +99%
BM_HpackParserParseHeader<NonIndexedBinaryElem<31, false>, UnrefHeader>          +10%        +10%
BM_HpackParserParseHeader<NonIndexedBinaryElem<31, true>, UnrefHeader>           +79%        +79%
BM_HpackParserParseHeader<NonIndexedElem, UnrefHeader>                           +42%        +42%
BM_HpackParserParseHeader<RepresentativeClientInitialMetadata, UnrefHeader>      +79%        +79%
BM_HpackParserParseHeader<RepresentativeServerInitialMetadata, UnrefHeader>      +68%        +67%
BM_HpackParserParseHeader<RepresentativeServerTrailingMetadata, UnrefHeader>     +146%       +146%
```

**Pull request merged**